includes/abstracts/abstract-wc-stripe-payment-gateway.php

```diff
@@ -1161,45 +1161,54 @@ public function process_refund( $order_id, $amount = null, $reason = '' ) {
 	 *
 	 * @since 3.0.0
 	 * @version 4.0.0
+	 *
+	 * @return array
 	 */
 	public function add_payment_method() {
-		$error     = false;
-		$error_msg = __( 'There was a problem adding the payment method.', 'woocommerce-gateway-stripe' );
-		$source_id = '';
+		try {
+			if ( ! is_user_logged_in() ) {
+				throw new WC_Stripe_Exception( 'No logged-in user found.' );
+			}
 
-		if ( empty( $_POST['stripe_source'] ) && empty( $_POST['stripe_token'] ) || ! is_user_logged_in() ) {
-			$error = true;
-		}
+			// Retrieve the source object from the submitted $_POST data.
+			$source_object = $this->get_source_object_from_request();
 
-		$stripe_customer = new WC_Stripe_Customer( get_current_user_id() );
+			if ( empty( $source_object ) || empty( $source_object->id ) ) {
+				throw new WC_Stripe_Exception( "The retrieved source doesn't contain an ID." );
+			}
 
-		$source = ! empty( $_POST['stripe_source'] ) ? wc_clean( wp_unslash( $_POST['stripe_source'] ) ) : '';
+			// Non-reusable payment methods won't be attached.
+			if ( ! WC_Stripe_Helper::is_reusable_payment_method( $source_object ) ) {
+				throw new WC_Stripe_Exception(
+					"The provided payment method isn't reausable." .
+					// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_print_r
+					PHP_EOL . 'Source object: ' . print_r( $source_object, true )
+				);
+			}
 
-		$source_object = WC_Stripe_API::get_payment_method( $source );
+			// Now that we've got the source object, attach it to the user.
+			$this->save_payment_method( $source_object );
 
-		if ( isset( $source_object ) ) {
-			if ( ! empty( $source_object->error ) ) {
-				$error = true;
-			}
+		} catch ( WC_Stripe_Exception $e ) {
+			$generic_error_msg = __( 'There was a problem adding the payment method.', 'woocommerce-gateway-stripe' );
 
-			$source_id = $source_object->id;
-		} elseif ( isset( $_POST['stripe_token'] ) ) {
-			$source_id = wc_clean( wp_unslash( $_POST['stripe_token'] ) );
-		}
+			WC_Stripe_Logger::log(
+				sprintf(
+					'Add payment method Error: %1$s %2$s%3$s',
+					$generic_error_msg,
+					PHP_EOL,
+					$e->getMessage()
+				)
+			);
 
-		$response = $stripe_customer->add_source( $source_id );
+			wc_add_notice( $generic_error_msg, 'error' );
```

Comment:
Question: Now that we're returning `[ 'result' => 'failure' ]` from this function, do we still need to add a notice with our own generic message?

WooCommerce Core outputs a generic message when the result is "failure" [here](https://github.com/woocommerce/woocommerce/blob/58dfbd2c4e305a2aa05a28c6ad9f5baeb29a2b6a/plugins/woocommerce/includes/class-wc-form-handler.php#L534) and while testing this PR, I noticed a bit of a double-up with two messages saying basically the same thing 😄 

![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/bc8db786-2f74-4628-99a5-959fc6052a8b)

I'm fine to keep it how it is currently since it's not a blocking issue.

Reply:
I fully agree. Thanks for bringing up that! Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2659/commits/ca7ab3f3a7ade935803d9c1a8602a01e35eaec97